crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,693 @@
+            )
+            | (
+                ModuleResolutionPathRef::StandardLibrary(StandardLibraryPath(self_path)),
+                ModuleResolutionPath::StandardLibrary(StandardLibraryPathBuf(other_path)),
+            )
+            | (
+                ModuleResolutionPathRef::SitePackages(SitePackagesPath(self_path)),
+                ModuleResolutionPath::SitePackages(SitePackagesPathBuf(other_path)),
+            ) => *self_path == **other_path,
+            _ => false,
+        }
+    }
+}
+
+impl<'a> PartialEq<ModuleResolutionPathRef<'a>> for ModuleResolutionPath {
+    fn eq(&self, other: &ModuleResolutionPathRef<'a>) -> bool {
+        other.eq(self)
+    }
+}
+
+impl<'a> PartialEq<FileSystemPath> for ModuleResolutionPathRef<'a> {
+    fn eq(&self, other: &FileSystemPath) -> bool {
+        self.as_file_system_path() == other
+    }
+}
+
+impl<'a> PartialEq<ModuleResolutionPathRef<'a>> for FileSystemPath {
+    fn eq(&self, other: &ModuleResolutionPathRef<'a>) -> bool {
+        self == other.as_file_system_path()
+    }
+}
+
+impl<'a> PartialEq<VfsPath> for ModuleResolutionPathRef<'a> {
+    fn eq(&self, other: &VfsPath) -> bool {
+        let VfsPath::FileSystem(other) = other else {
+            return false;
+        };
+        self.as_file_system_path() == &**other
+    }
+}
+
+impl<'a> PartialEq<ModuleResolutionPathRef<'a>> for VfsPath {
+    fn eq(&self, other: &ModuleResolutionPathRef<'a>) -> bool {
+        other.eq(self)
+    }
+}
+
+pub(crate) struct ModulePartIterator<'a> {
+    parent_components: Option<camino::Utf8Components<'a>>,
+    stem: Option<&'a str>,
+}
+
+impl<'a> ModulePartIterator<'a> {
+    #[must_use]
+    fn from_fs_path(path: &'a FileSystemPath) -> Self {
+        Self {
+            parent_components: path.parent().map(|path| path.components()),
+            stem: path.file_stem(),
+        }
+    }
+}
+
+impl<'a> Iterator for ModulePartIterator<'a> {
+    type Item = &'a str;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let ModulePartIterator {
+            parent_components,
+            stem,
+        } = self;
+
+        if let Some(ref mut components) = parent_components {
+            components
+                .next()
+                .map(|component| component.as_str())
+                .or_else(|| stem.take())
+        } else {
+            stem.take()
+        }
+    }
+
+    fn last(mut self) -> Option<Self::Item> {
+        self.next_back()
+    }
+}
+
+impl<'a> DoubleEndedIterator for ModulePartIterator<'a> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        let ModulePartIterator {
+            parent_components,
+            stem,
+        } = self;
+
+        if let Some(part) = stem.take() {
+            Some(part)
+        } else if let Some(components) = parent_components {
+            components.next_back().map(|component| component.as_str())
+        } else {
+            None
+        }
+    }
+}
```

Comment:
This is nead but doesn't seem needed at the moment. I recommend deleting it.

Reply:
:( Then we have to have the default implementation for `last()`, which exhausts the entire iterator to get the last element

Reply:
Oh, I wasn't aware that we use `last` somewhere.

Reply:
We don't. It just feels like a theoretical footgun to provide an `Iterator` interface that _could_ have a very efficient `last()` method, but doesn't.

Reply:
I would remove it then. I always assume that calling `last` on an iterator is `O(n)`. That's why I always double check if the iterator implements `DoubleEndedIterator` if I call the method in a hot code path.

Reply:
Alright :(